File: src/errata/models/anthropic.py
"""Anthropic (Claude) model adapter."""

from __future__ import annotations

import time
from collections.abc import Callable

from errata.models.base import AgentEvent, FileWrite, ModelAdapter, ModelResponse
from errata.tools import READ_TOOL_NAME, TOOL_DEFINITIONS, WRITE_TOOL_NAME, execute_read

# Translate canonical tool defs to Anthropic's input_schema format
_TOOLS = [
    {
        "name": t["name"],
        "description": t["description"],
        "input_schema": t["parameters"],
    }
    for t in TOOL_DEFINITIONS
]


class AnthropicAdapter(ModelAdapter):
    def __init__(self, model_id: str, api_key: str) -> None:
        self.model_id = model_id
        self._api_key = api_key

    async def run_agent(
        self,
        prompt: str,
        on_event: Callable[[AgentEvent], None],
        verbose: bool = False,
    ) -> ModelResponse:
        import anthropic

        client = anthropic.AsyncAnthropic(api_key=self._api_key)
        messages: list[dict] = [{"role": "user", "content": prompt}]
        proposed_writes: list[FileWrite] = []
        text_parts: list[str] = []
        start = time.monotonic()

        while True:
            response = await client.messages.create(
                model=self.model_id,
                max_tokens=8096,
                tools=_TOOLS,
                messages=messages,
            )

            tool_results = []
            for block in response.content:
                if block.type == "text":
                    text_parts.append(block.text)
                    if verbose:
                        on_event(AgentEvent("text", block.text))
                elif block.type == "tool_use":
                    if block.name == READ_TOOL_NAME:
                        path = block.input["path"]
                        on_event(AgentEvent("reading", path))
                        content = execute_read(path)
                        tool_results.append({
                            "type": "tool_result",
                            "tool_use_id": block.id,
                            "content": content,
                        })
                    elif block.name == WRITE_TOOL_NAME:
                        path = block.input["path"]
                        on_event(AgentEvent("writing", path))
                        proposed_writes.append(FileWrite(path, block.input["content"]))
                        tool_results.append({
                            "type": "tool_result",
                            "tool_use_id": block.id,
                            "content": "Write queued — will be applied if selected.",
                        })

            messages.append({"role": "assistant", "content": response.content})
            if tool_results:
                messages.append({"role": "user", "content": tool_results})

            if response.stop_reason == "end_turn":
                break

        return ModelResponse(
            model_id=self.model_id,
            text="".join(text_parts),
            latency_ms=int((time.monotonic() - start) * 1000),
            proposed_writes=proposed_writes,
        )

File: internal/models/anthropic.go
package models

import (
	"context"
	"encoding/json"
	"strings"
	"time"

	anthropic "github.com/anthropics/anthropic-sdk-go"
	"github.com/anthropics/anthropic-sdk-go/option"
	"github.com/suarezc/errata/internal/tools"
)

// AnthropicAdapter implements ModelAdapter for Anthropic Claude models.
type AnthropicAdapter struct {
	modelID string
	apiKey  string
}

// NewAnthropicAdapter creates an AnthropicAdapter.
func NewAnthropicAdapter(modelID, apiKey string) *AnthropicAdapter {
	return &AnthropicAdapter{modelID: modelID, apiKey: apiKey}
}

func (a *AnthropicAdapter) ID() string { return a.modelID }

func (a *AnthropicAdapter) RunAgent(
	ctx context.Context,
	prompt string,
	onEvent func(AgentEvent),
	verbose bool,
) (ModelResponse, error) {
	client := anthropic.NewClient(option.WithAPIKey(a.apiKey))

	toolParams := buildAnthropicTools()
	messages := []anthropic.MessageParam{
		anthropic.NewUserMessage(anthropic.NewTextBlock(prompt)),
	}

	var textParts []string
	var proposed []tools.FileWrite
	var totalInput, totalOutput int64
	var resolvedModel string
	start := time.Now()

	for {
		resp, err := client.Messages.New(ctx, anthropic.MessageNewParams{
			Model:     anthropic.Model(a.modelID),
			MaxTokens: 8096,
			Tools:     toolParams,
			Messages:  messages,
		})
		if err != nil {
			return ModelResponse{
				ModelID:      a.modelID,
				LatencyMS:    time.Since(start).Milliseconds(),
				InputTokens:  totalInput,
				OutputTokens: totalOutput,
				CostUSD:      CostUSD(a.modelID, totalInput, totalOutput),
				Error:        err.Error(),
			}, err
		}
		if resolvedModel == "" {
			resolvedModel = string(resp.Model)
		}
		totalInput += resp.Usage.InputTokens
		totalOutput += resp.Usage.OutputTokens

		// Append assistant turn
		messages = append(messages, resp.ToParam())

		var toolResults []anthropic.ContentBlockParamUnion
		for _, block := range resp.Content {
			switch block.Type {
			case "text":
				text := block.AsText().Text
				textParts = append(textParts, text)
				if verbose {
					onEvent(AgentEvent{Type: "text", Data: text})
				}

			case "tool_use":
				tu := block.AsToolUse()
				var input map[string]any
				_ = json.Unmarshal(tu.Input, &input)
				args := extractStringMap(input)

				switch tu.Name {
				case tools.ReadToolName:
					path := args["path"]
					onEvent(AgentEvent{Type: "reading", Data: path})
					content := tools.ExecuteRead(path)
					toolResults = append(toolResults, anthropic.NewToolResultBlock(tu.ID, content, false))

				case tools.WriteToolName:
					path := args["path"]
					onEvent(AgentEvent{Type: "writing", Data: path})
					proposed = append(proposed, tools.FileWrite{Path: path, Content: args["content"]})
					toolResults = append(toolResults, anthropic.NewToolResultBlock(tu.ID, "Write queued — will be applied if selected.", false))
				}
			}
		}

		if resp.StopReason == anthropic.StopReasonEndTurn || len(toolResults) == 0 {
			break
		}

		messages = append(messages, anthropic.NewUserMessage(toolResults...))
	}

	if resolvedModel == "" {
		resolvedModel = a.modelID
	}
	return ModelResponse{
		ModelID:        resolvedModel,
		Text:           join(textParts),
		LatencyMS:      time.Since(start).Milliseconds(),
		InputTokens:    totalInput,
		OutputTokens:   totalOutput,
		CostUSD:        CostUSD(a.modelID, totalInput, totalOutput),
		ProposedWrites: proposed,
	}, nil
}

func buildAnthropicTools() []anthropic.ToolUnionParam {
	var out []anthropic.ToolUnionParam
	for _, def := range tools.Definitions {
		props := map[string]any{}
		for name, p := range def.Properties {
			props[name] = map[string]any{
				"type":        p.Type,
				"description": p.Description,
			}
		}
		required := make([]string, len(def.Required))
		copy(required, def.Required)

		tp := anthropic.ToolParam{
			Name:        def.Name,
			Description: anthropic.String(def.Description),
			InputSchema: anthropic.ToolInputSchemaParam{
				Properties: props,
				Required:   required,
			},
		}
		out = append(out, anthropic.ToolUnionParam{OfTool: &tp})
	}
	return out
}

func extractStringMap(m map[string]any) map[string]string {
	out := make(map[string]string, len(m))
	for k, v := range m {
		if s, ok := v.(string); ok {
			out[k] = s
		}
	}
	return out
}

func join(parts []string) string {
	return strings.Join(parts, "")
}

func init() {
	// Ensure AnthropicAdapter satisfies ModelAdapter at compile time.
	var _ ModelAdapter = (*AnthropicAdapter)(nil)
}
